kube-core/src/params.rs

```diff
@@ -103,18 +103,22 @@ impl ListParams {
         }
         if let Some(continue_token) = &self.continue_token {
             qp.append_pair("continue", continue_token);
-        }
-
-        if let Some(rv) = &self.resource_version {
-            qp.append_pair("resourceVersion", rv.as_str());
-        }
-        match &self.version_match {
-            None => {}
-            Some(VersionMatch::NotOlderThan) => {
-                qp.append_pair("resourceVersionMatch", "NotOlderThan");
-            }
-            Some(VersionMatch::Exact) => {
-                qp.append_pair("resourceVersionMatch", "Exact");
+        } else {
+            // When there's a continue token, we don't want to set resourceVersion
+            if let Some(rv) = &self.resource_version {
+                if rv != "0" || (rv == "0" && self.limit.is_none()) {
+                    qp.append_pair("resourceVersion", rv.as_str());
+
+                    match &self.version_match {
+                        None => {}
+                        Some(VersionMatch::NotOlderThan) => {
+                            qp.append_pair("resourceVersionMatch", "NotOlderThan");
+                        }
+                        Some(VersionMatch::Exact) => {
+                            qp.append_pair("resourceVersionMatch", "Exact");
+                        }
+                    }
+                }
```

Comment:
doing the check here means we are making it the `ListParms` job to effectively lie about what it is doing and infer the users intent - which is not ideal for future maintenance because it gives a false impression of what listparams is doing.

the error case you are trying to catch here should be caught in `ListParams::validate` and returned (as opposed to silently truncating the params), and we should instead only set the version_match parameters in the necessary case in watcher. `Request::list` (used by `Api::list`) is already guarding on `ListParams::validate`.

Reply:
I get your point, however, if we move that to the validation step, this request `ListParams::default().limit(5).match_any()` is still technically valid, the server would process it normally, but the response may contain more than 5 items, which to me would look like a bug.

Unless we add this as validation as well, but it no longer match the validations we have on the apiserver (was it supposed to match?!) and it would shift some logic to the clients where they'd need to be aware of this behaviour

Reply:
hm, yeah, those are fair points. Thank you.

I think you are right. Particularly with the clear phrasing in the docs upstream:

> Setting the resourceVersionMatch parameter without setting resourceVersion is not valid.

we can justify this parameter truncation.

However, I do think this deserves a note about it in the `::match_any` builder docs (and for the other similar builders). Perhaps something to the effect of: "this option only affects cases where the resource version is set".